core/src/main/java/hudson/model/AbstractCIBase.java

```diff
@@ -192,19 +204,10 @@ private Computer createNewComputerForNode(Node n, boolean automaticAgentLaunch)
     }
 
     /*package*/ void removeComputer(final Computer computer) {
-        Queue.withLock(new Runnable() {
-            @Override
-            public void run() {
-                Map<Node, Computer> computers = getComputerMap();
-                for (Map.Entry<Node, Computer> e : computers.entrySet()) {
-                    if (e.getValue() == computer) {
-                        computers.remove(e.getKey());
-                        computer.onRemoved();
-                        return;
-                    }
-                }
-            }
-        });
+        final ConcurrentMap<Node, Computer> computers = getComputerMap();
+        if (computers.values().remove(computer)) {
```

Comment:
This changes the current behaviour where we now remove before locking the queue. We probably should leave this as locking before removing the entry.

This would result in code such as https://github.com/jenkinsci/jenkins/blob/c92096b7f5e735c924c5b5758f9166ade3d3a546/core/src/main/java/hudson/model/Queue.java#L1490 having a Computer that has been removed.